man/man1/vmup.1
.\" ############################################################################
.\" ### /usr/local/share/man/man1/vmup.1.gz BEGIN
.\" ############################################################################
.\"
.\" Copyright 2017 WhiteWinterWolf (www.whitewinterwolf.com)
.\"
.\" This file is part of vmtools.
.\"
.\" vmtools is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" ############################################################################
.
.Dd March 10, 2017
.Dt VMUP 1
.Os vmtools
.
.
.Sh NAME
.
.Nm vmup ,
.Nm vmcreate
.Nd start or create a Qemu virtual machine
.
.
.Sh SYNOPSIS
.
.Nm vmup
.Op Fl bhinqvyz
.Op Fl C Ar storage
.Op Fl c Ar storage
.Op Fl D Ar storage
.Op Fl d Ar storage
.Op Fl o Ar setting Ns = Ns Ar value
.Op Fl r | s
.Op Fl t Ar template_name
.Op Ar path ...
.Pp
.Nm vmcreate
.Op Fl bhinqvyz
.Op Fl C Ar storage
.Op Fl c Ar storage
.Op Fl D Ar storage
.Op Fl d Ar storage
.Op Fl o Ar setting Ns = Ns Ar value
.Op Fl r | s
.Op Fl t Ar template_name
.Op Ar vmhome ...
.
.
.Sh DESCRIPTION
.
.Ss Nm vmup
Boot each
.Ar path
as a Qemu virtual machine.
.Ar path
may be a path to a virtual machine home directory created using
.Nm vmcreate
or a storage path (see
.Sx Storage path
below).
.Pp
If no
.Ar path
is given,
.Nm vmup
assumes that the current working directory is a virtual machine home directory
created using
.Nm vmcreate.
.
.Ss Nm vmcreate
Create a new virtual machine in the directory
.Ar vmhome .
.Pp
If the directory does not already exists, it will be created.
If it already exists, it must not already contain a virtual machine.
If no path is given, the current working directory is assumed.
.
.
.Ss Storage path
.
The
.Ar storage
value may be one of the following:
.Bl -bullet -offset indent
.It
A path to a media image file natively supported by Qemu.
.br
This includes CD and DVD ISO images, raw and QCow2 hard-disk images.
.It
A path to a third-party or legacy hard-disk image file.
.br
.Nm vmcreate
and
.Nm vmup
propose to automatically convert such files into a QCow2 file, otherwise
such files are loaded in snapshot mode by default.
Conversion of
.Pa .ova
files is also supported (Qemu doesn't support them natively).
.It
A path to a local device file.
.It
A URL to a remote file or network storage.
.br
The protocols supported by Qemu are FTP(S), HTTP(S), iSCSI, NBD, SSH and TFTP.
Such files are loaded in snapshot mode by default, except iSCSI and NBD URLs
which are accessed in read-write mode by default.
See
.Xr qemu-system 1
to get more information regarding the URL syntax.
.It
A path to a directory.
.br
.Nm vmup
and
.Nm vmcreate
dynamically generates a device allowing to share the directory content with
the guest.
Limitations and requirements vary depending on the selected device type:
.Bl -dash -offset indent
.It
With hard-disk devices, Qemu virtual VFAT (VVFAT) feature is used.
The total directory size cannot exceed 504 MB and the host must not modify the
shared directory content while the guest is running.
The shared directory is accessed in read-only mode by default, but read-write
mode is also supported.
.Pp
This feature is not documented in the Qemu man page but is natively supported.
.It
With CD-ROM devices,
.Xr genisoimage 1
is used to automatically generate an ISO image storing the directory content.
If this command is not available on the host this feature is not available.
.\" TODO: Mention SMB share ("smb=dir" in Qemu doc) when using user mode net.
.El
.El
.Pp
Additionally, the storage path can be prefixed to explicitly set its access
mode:
.Pp
.Bl -tag -width "snap:" -offset indent -compact
.It Ic rw:
Read-write mode.
.It Ic snap:
Snapshot / non-persistent mode.
.It Ic ro:
Read-only mode.
.El
.Pp
The colon is part of the prefix (see
.Sx EXAMPLES
below).
Depending on the device settings and storage path used, not all modes are
always available.
When incompatible settings are used, starting the virtual machine fails with an
(hopefully explicit) error message.
.Pp
See also the Storage section in
.Xr vmtools.conf 5
for more details.
.
.
.Sh OPTIONS
.
The options are as follows:
.Bl -tag -width Ds
.
.It Fl b
Display the boot media selection menu. The exact behavior depends on the BIOS
used by Qemu.
.
.It Fl C Ar storage
Like
.Fl c
but also set the boot order to boot from a hard-disk.
.Pp
.Em Warning:
Using this option only enforces booting from a hard-disk drive as opposed to
another media type. It cannot ensure that the system will effectively boot from
.Ar storage
if it is not the guest's first or only hard-disk device.
.
.It Fl c Ar storage
Associate the guest's first free hard-disk drive to an image file.
.Ar storage
may point to a regular file, a block device, a directory or be a URL (see
.Sx Storage path
above).
.Pp
If
.Ar storage
points to a directory, it will be used to setup a virtual VFAT (VVFAT) share
with the guest. The total size of such share cannot exceed 504 MB.
.Em Attention:
it is important that the host does not attempt to modify the
content of this directory as long as the guest is running.
.
.It Fl D Ar storage
Like
.Fl d
but also set the boot order to boot from a CD-ROM.
.Pp
.Em Warning:
Using this option only enforces booting from a CD-ROM drive as opposed to
another media type. It cannot ensure that the system will boot from
.Ar storage
if it is not the guest's first or only CD-ROM device.
.
.It Fl d Ar storage
Associate the guest's first free CD-ROM drive to an image file  (see
.Sx Storage path
above).
.Ar storage
may points to a regular file, a block device, a directory or be a URL.
.Pp
If
.Ar storage
points to a directory, then
.Xr  genisoimage 1
must be available on the host and will be used to generate an ISO image
on-the-fly.
.
.It Fl h
Show usage information summary then exit.
.
.It Fl i
Interactive mode: do not daemonize,  when the virtual machine has been started
provide a Qemu monitor shell instead of exiting.
.Pp
When using this flag with
.Nm vmcreate ,
a final confirmation is requested after shutting down the virtual machine
allowing to quickly restart a guest installation in case of wrong manipulation.
.
.It Fl n
Do not start the virtual machine, all the other processing is done normally.
This allows to create a new virtual machine or convert some files without
booting them.
.
.It Fl o Ar setting Ns = Ns Ar value
Affect
.Ar value
to
.Ar setting ,
overriding any previous value set in
.Xr vmtools 7
configuration or virtual machine settings.
Available settings are listed in the
.Xr vmtools.conf 5
file.
.Pp
When used with
.Nm vmcreate ,
settings set using this option will also be stored in the newly created virtual
machine settings file.
.
.It Fl q
Decrease verbosity.
Add several
.Fl q
options to decrease verbosity even more
.Fl ( qq
by default to get minimal output).
See
.Xr vmtools.conf 5
for more information about verbosity levels.
.Pp
See
.Fl v
to increase the verbosity level.
.
.It Fl r
Do not modify storage backend files: enforce access to be read-only (write
attempts from the guest will fail).
Depending on the device settings, Qemu may refuse to start when read-only mode
is enforced.
.
.It Fl s
Do not modify storage backend files: use Qemu snapshot (non-persistent) mode
(the storage is seen as read-write by the guest, but modifications are stored
in a temporary file which is dropped once the virtual machine is stopped).
.Pp
This parameter has no effect on read-only storage.
.Pp
.Em Warning:
when manipulating sensitive data, ensure to always work on a copy instead of
the original file and/or unset the write bit on the backing file (see
.Xr chmod 1 ) .
Files with the write bit unset seem to be automatically loaded in snapshot mode
by Qemu.
.
.It Fl t Ar template
Apply a template,
.Ar template
must be a valid template name: the name of a
.Pa .inc.sh
file from the
.Pa templates
directory without its extension (see
.Sx FILES
below).
.Pp
When used with
.Nm vmcreate ,
template selected using this flag will also be added to the newly created
virtual machine settings file.
.
.It Fl v
Increase verbosity.
Add several
.Fl v
options to increase verbosity even more
.Fl ( vvv
by default to get the most verbose output, including debugging messages).
See
.Xr vmtools.conf 5
for more information about verbosity levels.
.Pp
See
.Fl q
to decrease the verbosity level.
.
.It Fl y
Don't ask any question: accept all confirmation requests and automatically
select the default answer in any other situation.
.Pp
Use this option with
.Em great care :
no confirmation will be asked before deleting or overwriting any files!
.
.It Fl z
Compress copied or converted disk image files.
This only affects QCow2 hard-disk images being copied or converted, data
written by a running guest is never compressed.
.El
.
.
.Sh ENVIRONMENT
.
.Bl -tag -width Ds
.It Ev TMPDIR
Default location to store temporary files, by default
.Pa /tmp .
.
.It Ev XDG_CONFIG_HOME
Location of user's configuration files, by default
.Pa ~/.config .
.El
.
.
.Sh FILES
.
.Bl -tag -width Ds
.It Pa /usr/local/lib/vmtools
Libraries shared by the vmtools project utilities.
.It Pa /usr/local/share/vmtools/modules/buildcmd
Modules used to build the Qemu command-line.
.It Pa /usr/local/share/vmtools/modules/configure
Modules used to select the virtual machine templates and define its settings.
.It Pa /usr/local/share/vmtools/templates
Virtual machines templates.
.It Pa /usr/local/share/vmtools/vmtools.conf
Virtual machine default settings, see
.Xr vmtools.conf 5 .
.El
.Pp
Moreover, the content of
.Pa /usr/local/share/vmtools
can be overridden in the following locations (in the order of precedence):
.Bl -tag -width Ds
.It ~/.config/vmtools
User overrides (if
.Cm cfg_include_userhome
is set to
.Dq yes ) .
.It /etc/vmtools
System-wide overrides.
.El
.
.
.Sh EXIT STATUS
.
.Bl -tag -compact -width Ds
.
.It 0
The operation ended successfully.
.It 1
An error occurred, incomplete tasks have been reverted.
.It 2
Wrong usage, nothing has been done.
.El
.
.
.Sh EXAMPLES
.
.Ss Nm vmup
.Bl -inset
.It Example 1:
Boot from a local ISO file:
.Pp
.Dl Ic vmup ./foo.iso
.Pp
.It Example 2:
Boot from a local hard-disk disk image, but in snapshot mode:
.Pp
.Dl Ic vmup snap:./foo.vmdk
.Pp
.It Example 3:
Boot from a remote URL:
.Pp
.Dl Ic vmup 'http://example.com/path/to/file.iso'
.Pp
.It Example 4:
Boot using SPICE display instead of the default one:
.Pp
.Dl Ic vmup -o vm_display_type=spice ./foo.iso
.Pp
.It Example 5:
Boot a virtual machine created using
.Nm vmcreate :
.Pp
.Dl Ic vmup ~/path/to/my_vm
.Pp
.It Example 6:
Boot the current virtual machine created using
.Nm vmcreate :
.Pp
.Bd -literal -offset indent
.Ic cd ~/path/to/my_vm
.Ic vmup
.Ed
.Pp
.It Example 7:
Boot a virtual machine created using
.Nm vmcreate ,
but start from a live CD instead of the main guest's disk:
.Pp
.Dl Ic vmup -D ~/path/to/livecd.iso ~/path/to/my_vm
.Pp
.It Example 8:
Convert a third-party image file into a QCow2 file without booting
it:
.Pp
.Dl Ic vmup -yn ./foo.ova
.Pp
.El
.
.Ss Nm vmcreate
.Bl -inset
.It Example 1:
Use an installation CD-ROM to create a new virtual machine:
.Pp
.Dl Ic vmcreate -d ~/path/to/install.iso ~/path/to/my_new_vm
.Pp
.It Example 2:
Use an installation CD-ROM to create a new virtual machine, and propose to
quickly restart the installation if needed:
.Pp
.Dl Ic vmcreate -i -d ~/path/to/install.iso ~/path/to/my_new_vm
.Pp
.It Example 3:
Use an already existing hard-disk image file to create a new virtual machine:
.Pp
.Dl Ic vmcreate -c ~/path/to/disk.ova ~/path/to/my_new_vm
.Pp
.It Example 4:
Use an already existing hard-disk image file to create three new virtual
machines using the settings given as parameters
.Fl ( o
options) and without requiring any interaction
.Fl ( y
flag):
.Pp
.Bd -literal -offset indent
.Ic vmcreate -o cpu_count=2 -o vm_ram_size=2G -y \e
.Ic "    " -c ~/path/to/disk.ova ./vm_1 ./vm_2 ./vm_3
.Ed
.Pp
.El
.
.
.Sh SEE ALSO
.
.Xr vmdown 1 ,
.Xr vmps 1 ,
.Xr vmtools 7
.
.
.Sh CAVEATS
.
.Bl -bullet
.It
Qemu VVFAT is not mentioned in Qemu man page, its implementation may
potentially be fragile.
.It
There is a slight delay in searching free TCP ports and opening them which can
lead to a race condition, specially when several virtual machines are started
in parallel, making some of them to fail to start.
.br
The recommended way to start several virtual machines at once is to pass them
as parameter to a single
.Nm vmup
command. This will also avoid any similar limitations affecting Qemu itself.
.El
.
.
.Sh REPORTING BUGS
.
Please send bug reports to the
.Lk http://github.com/WhiteWinterWolf/vmtools/issues "vmtools issues page" .
.
.\" ############################################################################
.\" ### /usr/local/share/man/man1/vmup.1.gz END
.\" ############################################################################


lib/vmup.inc.sh
################################################################################
### /usr/local/lib/vmtools/vmup.inc.sh BEGIN
################################################################################
#
# Copyright 2017 WhiteWinterWolf (www.whitewinterwolf.com)
#
# This file is part of vmtools.
#
# vmtools is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# ------------------------------------------------------------------------------
#
# This library provides functions to start a virtual machine.
#
# Public functions:
#   vmup_inputsettings
#         Interactively ask the user to input the virtual machine settings.
#   vmup_parseargs_args arg...
#         Command-line argument parsing for `vmup' and `vmcreate'.
#   vmup_runqemu
#         Start the Qemu hypervisor process.
#
################################################################################

################################################################################
# Global variables
################################################################################

# If set to "yes", `vmup_runqemu()' will build Qemu command-line and run it,
# otherwise this function will only output some diagnostic messages depending
# on the verbosity level and exit without starting Qemu.
vmup_runqemu_enable='yes'


################################################################################
# Functions
################################################################################

###
# vmup_inputsettings
#
# Interactively ask the user to input the virtual machine settings.
#
# The exact user interaction is controlled by the `configure/settings' modules.
#
vmup_inputsettings() {
	local 'mods_list' 'module'

	if [ "$cfg_ui_assumeyes" != 'yes' ]
	then
		echo "-----------------------------------------------------------------"
		echo "Answer to the questions below to define the guest properties."
		echo "You can type '?' to display the help screen at any time."
		echo "-----------------------------------------------------------------"
	fi

	# `settings_apply()' must be called before loading templates
	# to setup storage and after it to ensure that command-line settings
	# properly override anything else. Interactive settings input comes on top
	# of this.
	# See also `settings_loadvm()'.
	settings_apply || return 1
	mods_list=$( str_explode -- "$cfg_modules_configure_templates" ) || return 1
	for module in $mods_list
	do
		cli_trace 3 "vmup_inputsettings: ${vm_name}: configuring templates:" \
			"${module}."
		include_module -- "configure/templates/${module}.inc.sh" || return 1
		mod_configure || return 1
	done
	template_apply || return 1
	settings_apply || return 1

	mods_list=$( str_explode -- "$cfg_modules_configure_settings" ) || return 1
	for module in $mods_list
	do
		cli_trace 3 "vmup_inputsettings: ${vm_name}: configuring VM settings:" \
			"${module}."
		include_module -- "configure/settings/${module}.inc.sh" || return 1
		mod_configure || return 1
	done
}

###
# vmup_parseargs_args arg...
#
# Parse the command-line argument for the `vmup' and `vmcreate' commands.
#
# Sets the global variable OPTIND to allow shifting the parameters in the
# calling context (see http://pubs.opengroup.org/onlinepubs/9699919799/utilities/getopts.html).
#
# TODO: Add -a and -A with same behavior as -c/-C/-d/-D but with floppies
# instead of CD-ROMs (this will also require a proper template to support
# legacy platforms, maybe create `legacy90' and `legacy2k' templates with the
# presence of floppy drives trigerring the first one in the autodetect module).
#
vmup_parseargs() {
	# `$OPTIND' must not be available to the caller, do not set it local!
	local 'OPTARG' 'param' 'usage' 'writable_flag'

	if [ -z "${usage:-}" ]
	then
		echo "WARNING (BUG): \$usage undefined." >&2
		usage=""
	fi

	# Global variable preventing any forbidden combination of -r and -s flags.
	writable_flag=''

	OPTIND=1
	while getopts "bC:c:D:d:hino:qrst:vyz" param
	do
		case "$param" in
			'b') # Enable BIOS' boot media selection menu.
				settings_set 'vm_boot_menu' 'yes'
				;;

			'C') # Like `-c', but also set boot order to boot from the HDD.
				settings_setstorage 'hdd' "$OPTARG" || return 2
				settings_set 'vm_boot_order' 'c'
				;;

			'c') # Add a hard-disk drive image file.
				settings_setstorage "hdd" "$OPTARG" || return 2
				;;

			'D') # Like `-d', but also set boot order to boot from CD-ROM.
				settings_setstorage "cdrom" "$OPTARG" || return 2
				settings_set 'vm_boot_order' 'd'
				;;

			'd') # Add a CD-ROM image file.
				settings_setstorage "cdrom" "$OPTARG" || return 2
				;;

			'h') # Show usage information.
				printf '%s\n' "$usage"
				exit 0
				;;

			'i') # Interactive mode, do not daemonize but provide a QMP shell.
				settings_set 'vm_qemu_daemonize' 'no'
				;;

			'n') # Do not start the VM, only generate the configuration.
				vmup_runqemu_enable='no'
				;;

			'o') # Override a setting.
				value=$( str_unescape -- "${OPTARG#*=}" ) || return 2
				settings_override "${OPTARG%%=*}" "$value" || return 2
				;;

			'q') # Decrease verbosity.
				if [ "${cfg_ui_verbosity:?}" -gt 0 ]
				then
					cfg_ui_verbosity=$(( cfg_ui_verbosity - 1 )) || return 1
				fi
				;;

			'r') # Read-only access to all storage image files.
				if [ -z "$writable_flag" ]
				then
					writable_flag="r"
					settings_set 'vm_storage_rwmode' 'ro'
				elif [ "$writable_flag" != "r" ]
				then
					echo "ERROR: You cannot mix options such as 'read-only'" \
						"(-r) and 'snapshot-mode' (-s) together." >&2
					return 1
				fi
				;;

			's') # Do not modify storage (Qemu snapshot / non-persistent mode).
				if [ -z "$writable_flag" ]
				then
					writable_flag="s"
					settings_set 'vm_storage_rwmode' 'snap'
				elif [ "$writable_flag" != "s" ]
				then
					echo "ERROR: You cannot mix options such as 'read-only'" \
						"(-r) and 'snapshot' (-s) together." >&2
					return 1
				fi
				;;

			't') # Apply a template.
				template_add "$OPTARG"
				;;

			'v') # Increase verbosity.
				cfg_ui_verbosity=$(( ${cfg_ui_verbosity:?} + 1 )) || return 1
				;;

			'y') # Never ask any confirmation, always assume `y' as answer.
				settings_override 'cfg_ui_assumeyes' 'yes'
				;;

			'z') # Compress copied or converted images.
				settings_set 'vm_qemu_compress' 'yes'
				;;

			*)
				echo "Unexpected argument: $1" >&2
				return 2
				;;
		esac
	done

	settings_set 'cfg_ui_verbosity' "${cfg_ui_verbosity:?}"
	if [ "${cfg_ui_verbosity:?}" -ge 5 ]
	then
		set -x
	fi
}

###
# vmup_runqemu
#
# Start the Qemu hypervisor process, effectively booting the virtual machine.
#
# If the current verbosity level is greater or equal to 3, this function will
# also display on stderr the Qemu commmand-line which will be used.
# If the current verbosity level is greater or equal to 4, it will also dump
# current VM settings.
#
vmup_runqemu() {
	local 'cmd' 'mods_list' 'module'
	cmd=''

	if [ "${cfg_ui_verbosity:?}" -ge 4 ]
	then
		echo "*** START of VM settings dump:"
		set | awk '$0 ~ /^cfg_/ || $0 ~ /^vm_/ { print "    ",$0 }'
		echo "*** END of VM settings dump."
	fi

	mods_list=$( str_explode -- "$cfg_modules_buildcmd" ) || return 1
	for module in $mods_list
	do
		cli_trace 3 "vmup_runqemu: ${vm_name}: building Qemu command-line:" \
			"${module}."
		include_module -- "buildcmd/${module}.inc.sh" || return 1
		mod_buildcmd 'cmd' || return 1
	done

	cmd=$( str_escape -s ' ' -- $cmd )

	if [ "${cfg_ui_verbosity:?}" -ge 3 ]
	then
		printf '\nQemu command-line:\n'
		printf '%s\n' "$cmd"
	fi

	if [ "$vmup_runqemu_enable" = 'yes' ]
	then
		if [ -n "$vm_home" ]
		then
			lock_check -e || return 1

			# QEMU does not overwrite PID files correctly: it simply overwrites
			# the new PID over the current content of the file, which results
			# in a corrupted content when the new PID is shorter than the old
			# one currently stored in the file.
			# We therefore need to delete any PID file before starting QEMU.
			rm -f -- "${vm_home}/${cfg_file_pid:?}" || return 1

			# Some path may be relative to the VM home dir (monitor socket
			# file for instance).
			# TODO: Should the directory be placed earlier? No too early though
			# as path passed by the user as command-line parameter must remain
			# relative to the user's current directory.
			cd -- "$vm_home" || return 1
		fi
		eval "$cmd" || return 1
	fi
}

################################################################################
### /usr/local/lib/vmtools/vmup.inc.sh END
################################################################################
